kategory-effects/src/main/kotlin/kategory/effects/data/DeferredKW.kt

```diff
@@ -0,0 +1,110 @@
+package kategory
+
+import kategory.effects.data.internal.Platform
+import kategory.effects.instances.DeferredKWInstances
+import kotlinx.coroutines.experimental.*
+import kotlin.coroutines.experimental.CoroutineContext
+
+typealias DeferredResult<A> = Either<Throwable, A>
+
+@higherkind data class DeferredKW<out A>(val coroutineContext: CoroutineContext, val thunk: (CoroutineContext) -> Deferred<A>) : DeferredKWKind<A> {
+
+    fun runDeferred(): Deferred<A> =
+            thunk(coroutineContext)
+
+    fun <B> map(f: (A) -> B): DeferredKW<B> =
+            flatMap { a: A -> pure(coroutineContext, f(a)) }
+
+    fun <B> flatMap(f: (A) -> DeferredKW<B>): DeferredKW<B> =
+            DeferredKW(coroutineContext) {
+                this@DeferredKW.attempt().fold(
+                        { raiseError<B>(coroutineContext, it).runDeferred() },
+                        { f(it).runDeferred() })
+            }
+
+    companion object {
+        inline operator fun <A> invoke(coroutineContext: CoroutineContext, noinline f: suspend CoroutineScope.() -> A): DeferredKW<A> =
+                DeferredKW(coroutineContext) { context: CoroutineContext ->
+                    Platform.onceOnly<Unit> { Unit }.let {
+                        async(context, CoroutineStart.DEFAULT) {
+                            f()
+                        }
+                    }
+                }
+
+        inline fun <A> async(coroutineContext: CoroutineContext, noinline fa: Proc<A>): DeferredKW<A> =
+                DeferredKW(coroutineContext) { context: CoroutineContext ->
+                    Platform.onceOnly<Unit> { Unit }.let {
+                        async(context, CoroutineStart.DEFAULT) {
+                            IO.async(fa).unsafeRunSync()
+                        }
+                    }
+                }
+
+        inline fun <A> pure(coroutineContext: CoroutineContext, a: A): DeferredKW<A> =
```

Comment:
`pure` lifts an already computed value so it should not need to call `async` or need a context at all. I think we need here an actual ADT with `Pure`, `Async` and `Failed` because the context is only needed in async computations. that would remove the context from the constructor and reduce the overhead. Same applies for raising errors. I'm becoming convinced we don;t need the coroutine context at all except in the `async` convenience method. 
Furthermore if `bindAsync` already allows you to selectively use one in the middle of corutines and in bteween flatMaps the context can be entirely removed from this datatype.

Reply:
My concern is that you'd be doing work and for a pure or error you'll be reporting on the wrong thread.

For adoption it'll be important for people to control their threads, as in mobile the thread where you operate matter for several frameworks, like UI and Database.